test/test_sparse.py

```diff
@@ -2966,7 +2972,149 @@ def sparse_log(x):
 
+    @skipIfRocm
+    def test_sparse_matmul(self):
+        """
+        This function test `torch.sparse.mm` when both the mat1 and mat2 are sparse tensors. 
+        """
+
+        def _indices2csr(indices, dim):
+            nnz = len(indices)
+            r = [0] * (dim + 1)
+            last_i = 0
+            for i in indices:
+                if i != last_i:
+                    for _i in range(last_i, i + 1):
+                        r[_i + 1] = r[last_i + 1]
+                    last_i = i
+                r[last_i + 1] += 1
+            for _i in range(last_i, dim):
+                r[_i + 1] = r[last_i + 1]
+            assert r[-1] == nnz
+            return r
+
+        def sparse_mm(a, b, method='scipy'):
+            a = a.to('cpu')
+            b = b.to('cpu')
+            if method == 'scipy':
+                indices_1 = a._indices().numpy()
+                values_1 = a._values().numpy()
+                indices_2 = b._indices().numpy()
+                values_2 = b._values().numpy()
+
+                mat1 = scipy.sparse.coo_matrix((values_1, (indices_1[0], indices_1[1])), shape=a.shape)
+                mat2 = scipy.sparse.coo_matrix((values_2, (indices_2[0], indices_2[1])), shape=b.shape)
+                result = mat1.dot(mat2).tocoo()
+                return torch.sparse_coo_tensor([result.row, result.col], result.data, result.shape)
+            else:
+                assert a.shape[1] == b.shape[0]
+                n, p = a.shape
+                p, m = b.shape
+                indices_a = a._indices()
+                values_a = a._values()
+                indices_b = b._indices()
+                values_b = b._values()
+                nnz1 = len(indices_a[0])
+                nnz2 = len(indices_b[0])
+
+                if a.is_coalesced() and b.is_coalesced():
+                    r2 = _indices2csr(indices_b[0], b.shape[0])
+                    d = defaultdict(values_b.numpy().dtype.type)
+                    for n1 in range(nnz1):
+                        for n2 in range(r2[indices_a[1][n1]], r2[indices_a[1][n1] + 1]):
+                            d[indices_a[0][n1].item(), indices_b[1][n2].item()] += values_a[n1] * values_b[n2]
 
+                else:
+                    d = defaultdict(values_b.numpy().dtype.type)
+                    for n1 in range(nnz1):
+                        for n2 in range(nnz2):
+                            if indices_b[0][n2] == indices_a[1][n1]:
+                                d[indices_a[0][n1].item(), indices_b[1][n2].item()] += values_a[n1] * values_b[n2]
+                i3 = []
+                j3 = []
+                values = []
+                for i, j in sorted(d):
+                    i3.append(i)
+                    j3.append(j)
+                    values.append(d[i, j])
+                return torch.sparse_coo_tensor(torch.tensor([i3, j3]), torch.tensor(values), (n, m))
+
+        def test_grad_with_custom_sparsity_pattern(sparse_dims, nnz, shape_a, shape_b):
```

Comment:
Instead of starting these internal helpers with "test", start their names with "_helper" or something like it to clarify these aren't actual test cases